### ISSUE TITLE
Null-safety when using externally specified code

### ISSUE THREAD

**Issue by drpond1** [dart-lang/language#1531]
I am encountering an issue when I use code that is supposed to conform to an external spec.   To make things null-safe I have to 
introduce extra items in the code which make it harder to follow.

Example 1:
   In the DART sdk for the web audio api,   the typical return value of an external function is declared nullable,  when the W3C specification says otherwise.  For example,   GainNode.gain is typed by DART as nullable.

Example 2:
   When parsing a legal musicXML file,   a value is often known to be set before use because the spec requires a certain ordering of the XML elements in some sequence.    

......

What is happening as I try to update my code,  is that my code is becoming less clear because the types I am using don't match the types I am expecting from reading the specs.

For example 1, I would rather give up a possible null-safety  violation than have the code turn out this way.   My code is intended for the Chrome browser and with a high probability I trust that code.
I use with it within a try-catch so if the web-audio graph construction
 throws an unexpected exception I can recover fairly gracefully.
Not having the code reflect the spec is not worth the added safety check.

For example2,  there is a somewhat similar problem.   Only in this case
  there may indeed be a null problem, if the musicXML is not legal.   This is far more likely to happen than in example 1.  However, this translation is protected by a try-catch and I recover gracefully if the file does not parse and the internal data structure cannot be built.  Even if the problem shows up as an unexpected derived text string rather than an outright failure,  it would be ok.   The spec is sufficiently complicated that i would prefer a  null safety problem,  rather than having the misleading statements in the code.   
 
....
More Detail:

 The first example is not too bad,  quite a few extra variables that
   only serve to convince the compiler that something is not null.
 Well GainNode.gain is readOnly and is really always non-null.
     Instead of GainNode.gain   I use   
          AudioParam  gainParam=GainNode.gain!;
                 gainParam.setValue ......
     
 I could have used GainNode.gain! all over the place.

 In the second example,  I sometimes put in useless assignments when
      possibly "defined before use"  appears uncertain but can't happen due to the order of the expected XML elements in some method (and there is no danger of the initial value being propagated for correct 
  files).
         I might change what I am doing to something like.
           pitchedNoteAsString =  /* unused */ "c";  
       Which indicates what is supposed happen for legal files
    and the fall back for  bad files.  

......

But perhaps there are  better solutions  ???

**Comment by tatumizer** [comment]
Maybe you are looking for this?
```dart
late String pitchedNoteAsString;
```
It can also be "late final", with (lazily evaluated) initializer (default value) or without.
No?

**Comment by drpond1** [comment]
Yes,   late String pitchedNoteAsString is an option but indicates that the desired "backup" handling is to throw an exception.  

Here, I know of a default value that is less likely to cause a problem if the file is bad.    It is just not the value for good files.
(There was also no unexpected propagation of the default value. )

**Comment by tatumizer** [comment]
Can you write **in pseudocode** the logic you would like to have there? Or just in plain English. Otherwise, it's all very informal, not sure I understand your use case.

**Comment by drpond1** [comment]
I reread the description of the lazy evaluator for late and it doesn't seem right.   But I am not sure.

with your proposed scheme

 late String   pitchedNote = "c"

 For the good case the first thing happens is that
    pitchedNote = correctString;
    (pitchedNote is not "accessed"  here.  a setter not a getter is used.)
The second thing that happens in the good case
 is that pitched Note is accessed.
   so pitchedNote becomes "c" and then
   foo=pitchNote+alteration;

    so I get foo is  "c"+alteration


But what I want is
     pitchedNote=correctString
     foo=pitchedNote+alteration    

      foo becomes correctNote+alteration.


In the bad case
     foo=pitchedNote+alteration  is
      "c"+alteration   which is fine
 

  But perhaps I simply didn't understand the definition of
     "is accessed".  Perhaps the setter is meant to be included in 
     gets accessed.

Notice that my solution will produce the desired result.

**Comment by leafpetersen** [comment]
@drpond1 I'm not exactly sure what you're asking for here.  I think your first issue is around the typing of the html libraries?  @srujzs may be able to comment there.  I would need a more specific concrete description of the code you're trying to write in your second case to comment there.

**Comment by tatumizer** [comment]
As soon as you say `pitchedNote = correctString;`, forget about "c". Pitched note will never become "c". This "c" is the default value. It comes into play if you **read** the variable **before you ever write** to it explicitly. Does this answer your question?

**Comment by srujzs** [comment]
>Example 1:
>In the DART sdk for the web audio api, the typical return value of an external function is declared nullable, when the W3C specification says otherwise. For example, GainNode.gain is typed by DART as nullable.

We look at compatibility as well as nullability for the web libraries. We use the MDN compatibility data to determine whether a particular API is compatible across all browsers and versions (and sometimes, even platforms) we intend to support. For your example, https://developer.mozilla.org/en-US/docs/Web/API/GainNode indicates that `gain` is unsupported on IE, so a call to this getter will return `undefined` which gets translated to `null`. If the Dart type is non-nullable, this will break soundness silently. This does mean, however, that we play it "safe" in determining nullability of native APIs, which means that more APIs will end up being typed as nullable. There are more corner cases here as well, like unparseable information in the compatibility data, but this is why you'll see APIs that should have been typed as non-nullable being typed as nullable.

**Comment by drpond1** [comment]
To Leaf: 
I am not asking for any language changes.   I would simply like to write clearer code.     
1.  I suppose I first need to understand why the DART library has an interface that is different from the W3C spec.   is it because you can't trust the javascript interface to make things non-null?

2.  Here is a further simplification of the second example
     The steps to be performed depend on the structure of the
       XML file being parsed.    The children of the node being parsed
       are supposed to be ordered in a particular way.
       When I see the first element I assign
           pitchedNote a value say 5.    
      When I see the second element I use the value from the first step
            foo=pitchedNote+2;

      Now a good file will conform to the spec and hence I am guaranteed that pitchNote will be initialized before use.   However,
   there can be bad files.  So how to define pitchedNote in the code?

    I suggested:
         int pitchedNote= /*unused */ 1;
   
   This indicates by comment that pitchedNote does not need to
        be assigned for good files and that the value assigned will
        typically not be used.
   It also says the value 1 may be used if the file does not conform
        to the spec.
      

   a.  late pitchNote   is not an ideal because it throws when a bad file is found.
  b.   int pitchNote = 1  is not ideal,  because it suggests that the initial value may be used,  when that is not the normal behavior.
  c.  int? pitchNote   is not ideal because it suggests pitchNote might not  be specified with a good file.

 d.  tatumizer  has suggested
      late int pitchNote =1.
   If this works the way he says it does,  this would be better than 
     what I suggested, because the behavior does not need  
   comments to be explained.    

  
      be specified with a correct file.

**Comment by leafpetersen** [comment]
@drpond1 I'm sorry, I really don't understand what you are asking for, maybe a small fragment of actual code would help?  In particular, you write:

> a. late pitchNote is not an ideal because it throws when a bad file is found.
> b. int pitchNote = 1 is not ideal, because it suggests that the initial value may be used, when that is not the normal behavior.
> c. int? pitchNote is not ideal because it suggests pitchNote might not be specified with a good file.

So summarizing this, you don't want it to throw if it gets read before writing, and you don't want it to return a default value if it gets read before writing, and you don't want it to return `null` if it gets read before writing.  So... what do you want to happen?  I don't see a lot of options left.  One possible way of refactoring your API might be something like this:

```dart
class Example {
  int? _pitchNote;
  int get pitchNote => !_pitchNote;
  void set pitchNote(int value) { _pitchNote = value; }
  bool hasPitchNote => _pitchNote != null;
}
```

If that pattern doesn't do what you want, perhaps you can show me the behavior you are looking for with a similar example and I can help?  As it is though, I'm really not sure what behavior you are asking for.

**Comment by drpond1** [comment]
What I want is   what   tatumizer  has suggested.
late int pitchNote =1.

...  provided the "is accessed"  conforms to his definition 

  If pitchedNote gets set before being used ignore the initialization.
  If it hasn't set before its use then use the initialization before proceeding with the evaluation.  (If the initialization produces null,
   throw a null pointer exception.)

**Comment by leafpetersen** [comment]
There is no observable semantic difference between `late int pitchNote = 1` and `int pitchNote = 1`, so I don't really understand what you're aiming for.  Making a field with an initializer `late` makes the initializer lazy evaluated (and hence it can access `this` in the initializer) but there is no difference between lazily and eagerly evaluating `1`.

**Comment by drpond1** [comment]
I didn't think lazy evaluation with late would work either.   

However,  you can look at what Tatumizer has suggested to understand what I want.

**Comment by drpond1** [comment]
So in short you want to be able say
    
 do implicit null checking and if you encounter null from an attempted get,  then set the value to the default before continuing.

    Even better this would permit you to do all sorts of things.
         
       late c = () {print("error initializing c");  throw( someError); }
        late c  = (){ setup error information;   throw(someError); }
         late c  = (){  print("c failed to get initialized properly"); return(c)}
         late c = (){ d=value;  return(c);}
         late c=  (){ if (handlingLateC) {return default} else {throw(error); }
   
 So either the documentation might be improved OR there is a suggestion for additional functionality.

**Comment by tatumizer** [comment]
@drpond1: you can get all this functionality right now - just add parentheses to the expressions.
 late c = () {print("error initializing c");  throw( someError); }**()**;
late c  = (){ setup error information;   throw(someError); }**()**;
late c  = (){  print("c failed to get initialized properly"); return(c)}**()**;
late c = (){ d=value;  return(default);}**()**;
late c=  (){ if (handlingLateC) return default; else throw(error);}**()**;

**Comment by drpond1** [comment]
Sorry about the syntax error.   I upated my comment.

**Comment by drpond1** [comment]
Tatumizer must be correct, because it is the interpretation that makes sense.  

So my updated suggestion is to update the documentation with
  an explanation of the simple case of
      late int yummy= 24;

and a mention of the possibility of providing different error handling
    options when null is encountered when attempting to get a value for
    an uninitialized variable.

**Comment by drpond1** [comment]
I understand that there may be reasons not to have DART conform to a spec.

Still there is something not quite right when the documentation is so different from the spec.   For example  SelectionElement in HTML has an attribute selectedIndex that the spec says  returns  -1 if nothing is selected.
    the DART documentation only says the return type is int?  with no other
      information.   A casual reader would probably infer that null is 
     returned when nothing is selected.    Granted,  the DART spec makes it clear that it doesn't document everything.   But this simply seems too misleading.

**Comment by srujzs** [comment]
I can sympathize with it being misleading. Unfortunately, here are the options we have that I see:

1. Increase potential unsoundness by using nullability only and not compatibility. This is a no-go in my opinion, because it will (and has for the Flutter folks) result in all kinds of null-related errors. While it may be more intuitive for some users, it'll make it unusable for others.
2. Keep it as is, which has its own potential downsides, one of which you mentioned.
3. Add extra information via annotations like we've done in the past with `@SupportedBrowser` along with option 2. This has the possible upside of being informative but introduces _significant_ extra code, isn't exhaustive since it doesn't capture some other attributes that might exist (like platform compatibility or particular corner-cases), and will make it even more complex in terms of maintenance and accuracy. It doesn't seem like it will be more useful (or correct) than a user looking up the corresponding spec.

I see option 2 as the best compromise option. It unfortunately means that extra null checks might be needed but it is more robust for all browsers and minimizes what it promises.

**Comment by drpond1** [comment]
I think your response came before I updated my letter.  Thank you for the quick reply.   

If changing the type checking is problematic,  you could still change the documentation to something like I suggested.    I prefer to be able to use the DART documentation since it shows up while I am editing the code.     That change would avoid the specific issue I described.     (I can see that there is an issue about cluttering up the language or perhaps just the documentation part of the language.   So I can see why someone might have a different point of view.)

**Comment by drpond1** [comment]
More about ?!

  int?! x as an instance of a class in the interface could be combined with a
        setter that is only specified abstractly.  (Perhaps this is even done
automatically with ?! definitions applied to instances. )  The setter cannot be assumed to be simple assignment by the type checker and so the checker can never be sure it is ever non-null.   
 And you should not be allowed to assign an ?! instance a null value,  which might break the normal usage. 
  But unlike late,   you can say   int z=foo.x ?? 23;

**Comment by lrhn** [comment]
Sounds like you want a getter with type `int?` and a setter with parameter type `int`.
We don't currently allow that (we allow the other direction, though).
This is a seemingly valid use-case for that particular combination. It won't allow you to do `x = x;` (but who needs that anyway?)

I'm not sure I'd recommend allowing it. I'd rather disallow the other direction too, and require getters and setters to always have the same type.
What you can do is use a setter function *different* from the getter name:
```dart
int? get x;
void initializeX(int value) { x = value; }
```

About `?!` as a "nullable getter/non-nullable setter" declaration, I think it's aiming too low. If we'd allow different types for the getter and setter, in that particular direction, we should allow the getter to be *any* supertype of the setter, not just the nullable version of the same type.

**Comment by drpond1** [comment]
I know the desirability of having some way to see the underlying spec,  which was the original motivation for the ?! idea.   But beyond that I am not sure.   So I would be satisfied if the
?!  was handled identically to the functionality of ?.   There does not have to be a tighter restriction on the setter.    

In reality,  the specifications are often not well presented by any typing.
For example,  the height of a canvas in HTML is supposed to be a valid non-negative integer.
 (an unsigned long).
But,  it is permitted for a user to set the height to something that is not even parsable.
  If there is a parsing error,
  the browser is supposed to use the default value to 150.   So the type of the argument for the setter for height  is what?  What should DART check for?   The DART spec says  int?.     
 But if null is permitted,  why not a floating value?   If one says  ?! int,  then one gets the idea that
  it supposed to be an int.    But,  it is a mess no matter what is done.  


  In considering your generalization,  I am not sure how useful it would be.
    For example,  suppose  the getter-type was num and the setter type was int.
   If  the item was really supposed to be an int,  what would ever cause it to have a float value?
   Perhaps it is the result of importing some foreign object without having fully processed it.
      So something might temporarily be a num, but the intent is that it will be converted into an 
     int.      Perhaps the reason the processing is deferred is that rounding is not always the
    desired solution.     So a use case is plausible.   Yet,  I am not convinced that this kind of
   typing is providing much value.
